Add IP Prefix Aggregation-based Visualization

ION28 opened the pull request:
## Summary

Closes #156121

This PR is for Issue #156121 and adds the ability to perform [IP Prefix](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-ipprefix-aggregation.html) aggregation-based visualizations within the Kibana UI. Previously this aggregation could only be done in DevTools as a manual query to Elasticsearch and not visualized.

![image](https://github.com/elastic/kibana/assets/3931697/7e049da9-fd42-41f6-bbaf-99d3d6cb0790)

![image](https://github.com/elastic/kibana/assets/3931697/dcfae012-9d06-4346-9118-3965434ff8b8)


### Various Notes

* The following two folders & their subfolders had files modified for this
  * src/plugins/vis_default_editor/public/components
  * src/plugins/data/common/search/aggs
* I spent a fair amount of time debating & attempting to build the PrefixLength Input boxes and their interplay with the is_ipv6 toggle button. Originally I tried having only 1 PrefixLength button that the toggle switch would modify the max value / validate the contents of. 
  * In the end, it seemed much cleaner & straightforward to have two separate input boxes (both prefix_length.tsx components) and just create them with different options. This means that when a user toggles the switch back and forth, they would be seeing/editing two different Prefix Length boxes depending on which way the switch is. 
  * To make it a little more clear they are different boxes, I put "IPv4" and "IPv6" in the label name for these boxes. Additionally, I think it is helpful this way if you are potentially swapping back and forth between v4 and v6 visualizations.
* There is 4 new unit tests, all related to input options, added in the ip_prefix_fn.test.ts file
* Note - here is a test CSV file of IPv4 addresses one could import to test locally and see this addition. 
[alphadataset.csv](https://github.com/elastic/kibana/files/13691358/alphadataset.csv)
  * Configure the Override settings in this way if uploading it to Kibana/Elastic
  * 
![image](https://github.com/elastic/kibana/assets/3931697/34ce701a-f4d5-4107-8a08-c6195e21c169)


 
### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### Risk Matrix

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|



### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

ION28 commented:
CLA has been signed now.

elasticmachine commented:
Pinging @elastic/kibana-visualizations (Team:Visualizations)

stratoula commented:
/ci

stratoula commented on src/plugins/data/common/search/aggs/buckets/ip_prefix_fn.ts:
```diff
@@ -0,0 +1,107 @@
+        defaultMessage: 'Length of the network prefix',
+      }),
+    },
+    prefixLength64: {
```
Do we need two different options here? `prefixLength` and `prefixLength64`? Maybe the isIpv6 option can be used for the checks you need?

ION28 replied:
Updated to just have one variable that implements the approach you suggested

stratoula commented on src/plugins/data/common/search/aggs/buckets/create_filter/ip_prefix.ts:
```diff
@@ -0,0 +1,17 @@
+import { IBucketAggConfig } from '../bucket_agg_type';
+
+export const createFilterIpPrefix = (aggConfig: IBucketAggConfig) => {
+  return buildRangeFilter(
```
This doesn't seem right to me. Why rangeFilter here? I think this is causing the bugs with the filters creation. It would be awesome if we could add a test too!

ION28 replied:
Updated to use a proper Range Filter that converts the input into a CidrMask (like the ip_range filter) to resolve the behavior issue in your comment above

stratoula commented on src/plugins/vis_default_editor/public/components/agg_params_map.ts:
```diff
@@ -31,6 +31,11 @@ const buckets = {
   },
+  [BUCKET_TYPES.IP_PREFIX]: {
+    prefixLength: controls.PrefixLengthParamEditor,
+    prefixLength64: controls.PrefixLengthParamEditor,
```
As I said above I think one prefixLength parameter is enough here

ION28 replied:
Updated to just have one prefixLength input box

stratoula reviewed (commented):
Hey, thanx a lot for your contribution! Some comments from my side:

1. Clicking on the chart results to an error page. We should fix this by creating either the correct filter or by disabling the creation of this filter for this aggregation. I think the first is the best path forward
2. Same thing happens when I add the new aggregation on the breakdown and try to filter in/out from the legend 
<img width="2498" alt="image" src="https://github.com/elastic/kibana/assets/17003240/dc4c13f7-e99c-42f4-acd2-aa0d3fc76d58">

ION28 commented:
Okay thank you for all of the feedback @stratoula! I'll work on these fixes & let you know when they are ready for review again

ION28 commented:
Still working on a couple more items to make this PR ready for review again:
- [ ] Fixing Range Filter
- [ ] Adding Tests/Fixing CI issues

ION28 commented:
@elasticmachine merge upstream

ION28 commented:
/ci

ION28 commented:
Guess I can't trigger the CI 😅 Also seems like I can't access any of the Buildkite urls as a non-elastic employee; however, I did run the tests for the data plugin locally and all seems to pass. Will see what the CI says.

Couple notes now that I think this is ready for review again:
* I switched to using a single input box for the prefixLength. Because of how that and the isIpv6 switch behavior, I implemented both inputs within the same React Controlled Component. That way, changing the Switch from IPv4 to IPv6 and vice versa can dynamically change the Max value of the prefix Length input box.
* Given that setup with the values in the single component, I added a new type "ip_prefix" and the associated code to handle that in `src/plugins/data/common/search/expressions` 
* Since last review, I corrected the RangeFilter to work appropriately with the IP Addresses. I expect the most common use of the filter would be to filter out a particular IP Range from the resulting visual.
* Since last review, I also added tests for each of the components to match similar the unit tests similar components (e.g. IP Range or the Extended Bounds subcomponent) have

ION28 commented:
May need to also update this page of the docs https://github.com/elastic/kibana/blob/main/docs%2Fuser%2Fdashboard%2Fcreate-panels-with-editors.asciidoc

stratoula commented:
/ci

sabarasaba reviewed (approved):
deployment management changes lgtm

stratoula commented:
/ci

stratoula commented:
@ION28 I see the following errors in the CI

<img width="1040" alt="image" src="https://github.com/elastic/kibana/assets/17003240/1ef32938-2fe3-4ee8-926a-0aed4c5d33c6">

stratoula commented on src/plugins/data/common/search/aggs/buckets/create_filter/ip_prefix.ts:
```diff
@@ -0,0 +1,24 @@
+export const createFilterIpPrefix = (aggConfig: IBucketAggConfig, key: IpPrefixKey) => {
+  let range: RangeFilterParams;
+
+  range = new CidrMask(key.address + "/" + key.prefix_length).getRange();
```
Why are you using let here, you dont reassign it. 

```
const range = new CidrMask(key.address + "/" + key.prefix_length).getRange();
```

ION28 replied:
Fixed

stratoula commented on src/plugins/data/common/search/aggs/buckets/ip_prefix.ts:
```diff
@@ -0,0 +1,73 @@
+import { BucketAggType } from './bucket_agg_type';
+import { BUCKET_TYPES } from './bucket_agg_types';
+import { createFilterIpPrefix } from './create_filter/ip_prefix';
+import { IpPrefixKey, IpPrefixAggKey } from './lib/ip_prefix';
```
IpPrefixAggKey is not used

ION28 replied:
Fixed

stratoula commented on src/plugins/data/common/search/aggs/buckets/lib/ip_prefix.ts:
```diff
@@ -0,0 +1,19 @@
+
+export type IpPrefixKey = IpPrefixAggKey;
+
+export const convertIPPrefixToString = (range: IpPrefixKey, format: (val: any) => string) => {
```
range doesnt seem as the correct variable name here

ION28 replied:
Thanks for the feedback! I spent some time debating this more after your comment. Since the variable is of type "IpPrefixKey" which is basically an IP Address (e.g. 192.168.0.0) coupled with a Numeric Prefix (e.g /16) -> so 192.168.0.0/16 - I guess "cidr" would be a more appropriate name than range. This naming then ends up mirroring the naming in search/aggs/buckets/lib/cidr_mask.ts which also handles strings of the form "address/prefix" (e.g. 192.168.0.0/16).

I've updated the use to cidr here and in the next comment.

stratoula commented on src/plugins/data/common/search/aggs/utils/get_aggs_formats.ts:
```diff
@@ -113,6 +114,20 @@ export function getAggsFormats(getFieldFormat: GetFieldFormat): FieldFormatInsta
+      static id = 'ip_prefix';
+      static hidden = true;
+
+      textConvert = (range: IpPrefixKey) => {
```
Same here, why is it named range?

ION28 replied:
Fixed, see comment above

stratoula commented on src/plugins/data/common/search/expressions/ip_prefix.ts:
```diff
@@ -0,0 +1,55 @@
+  type: 'ip_prefix',
+  inputTypes: ['null'],
+  help: i18n.translate('data.search.functions.ipPrefix.help', {
+    defaultMessage: 'Create extended bounds',
```
These help messages seem wrong to me

ION28 replied:
Whoops, thank you for catching that, fixed to use language from https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-ipprefix-aggregation.html#search-aggregations-bucket-ipprefix-aggregation

stratoula commented on src/plugins/vis_default_editor/public/components/controls/ip_prefix.tsx:
```diff
@@ -0,0 +1,124 @@
+
+  if (!isValid) {
+    error = i18n.translate('visDefaultEditor.controls.ipPrefix.errorMessage', {
+      defaultMessage: 'Prefix Length must be between 0 and 32 for IPv4 Addresses and 0 and 128 for IPv6 Addresses.',
```
```suggestion
      defaultMessage: 'Prefix length must be between 0 and 32 for IPv4 addresses and 0 and 128 for IPv6 addresses.',
```

ION28 replied:
Fixed!

stratoula commented on src/plugins/vis_default_editor/public/components/controls/ip_prefix.tsx:
```diff
@@ -0,0 +1,124 @@
+}
+
+const prefixLengthLabel = i18n.translate('visDefaultEditor.controls.IpPrefix.prefixLength', {
+  defaultMessage: 'Prefix Length',
```
```suggestion
  defaultMessage: 'Prefix length',
```

ION28 replied:
Fixed!

stratoula commented on src/plugins/vis_default_editor/public/components/controls/ip_prefix.tsx:
```diff
@@ -0,0 +1,124 @@
+
+import React, { useCallback } from 'react';
+
+import { isUndefined } from 'lodash';
```
This doesnt seem to be used

ION28 replied:
Fixed

stratoula reviewed (commented):
Thanx for your effort, we are getting closer! I added some comments and some bugs I found

stratoula commented:
If I have an ipv6 split the filtering doesnt work and I get this error in the console

<img width="605" alt="image" src="https://github.com/elastic/kibana/assets/17003240/840093dc-c7c0-4744-901f-48a11298011e">


This is my configuration
<img width="1744" alt="image" src="https://github.com/elastic/kibana/assets/17003240/6991a3cb-e14f-4e93-804c-e9f406f9f1bf">

ION28 commented:
@elasticmachine merge upstream

ION28 commented:
@stratoula thank you for all of your feedback so far on the PR. I believe I have addressed all of your last round of comments. A couple notes:
* The explanation for the bug in your screenshot is described in [src/plugins/data/common/search/aggs/buckets/create_filter/ip_prefix.ts](https://github.com/elastic/kibana/pull/173474/files#diff-9bef687e47d647883a1ba54cecdd72b8974e5175caa3ca252a2939b4214c2c46) Given the complexity of when/how that scenario occurs and how I addressed it, I thought it was best to write it out as a block comment for anyone who might review it in the future.
* There is one additional case of potential weirdness I discovered during this last round of testing, and I can't think of any way to handle it. Your screenshot above is basically the opposite of this case. For this one, let's say you have an index/data view where one field only contains IPv6 addresses. You do a split series (or X-axis) on that field with a low prefix (say 2), and the is_ipv6 switch is set to "false." The keys in the legend (or the bars themselves) that you can filter on will show IPv4 addresses. However, clicking on them to either filter in or out will have no effect on the resulting chart. Unlike with the former case, I don't know how we would be able to detect that there are IPv6 addresses making up these bars and filter appropriately. I don't believe that Kibana has enough information in this scenario to know to map the IPv4s to IPv6s to filter the data. Not sure how you want to handle this behavior.
* I also added the small bit to update the documentation to specify support in Kibana for the IP Prefix aggregation.

stratoula commented:
@ION28 thanx a ton! I am running the CI and I will do another round of review/test in a while.

stratoula commented:
/ci

stratoula commented on src/plugins/data/common/search/aggs/buckets/create_filter/ip_prefix.ts:
```diff
@@ -0,0 +1,37 @@
+ * Side Public License, v 1.
+ */
+
+import { buildRangeFilter, RangeFilterParams } from '@kbn/es-query';
```
RangeFilterParams is not used and CI complains :D

ION28 replied:
Fixed

stratoula commented:
> There is one additional case of potential weirdness I discovered during this last round of testing, and I can't think of any way to handle it. Your screenshot above is basically the opposite of this case. For this one, let's say you have an index/data view where one field only contains IPv6 addresses. You do a split series (or X-axis) on that field with a low prefix (say 2), and the is_ipv6 switch is set to "false." The keys in the legend (or the bars themselves) that you can filter on will show IPv4 addresses. However, clicking on them to either filter in or out will have no effect on the resulting chart. Unlike with the former case, I don't know how we would be able to detect that there are IPv6 addresses making up these bars and filter appropriately. This "problem" occurs because Elasticsearch will auto-convert IPv6 addresses to IPv4s in the backend when doing the ip_prefix aggregation, thus IPv6 addresses end up being counted in buckets with an IPv4 address label in the response from ES. I don't believe that Kibana has enough information in this scenario to know to map the IPv4s to IPv6s to let the user the filter the data falling into these ranges out (unless both an IPv4 and an IPv6 filter got created every time the user clicks to apply a single filter). Not sure how you want to handle this behavior.

I think it is fine for now, I feel that this is an edge case. We will see how users interact with it and if they fall in this problem often and we will figure out something. I really hope that it is going to be used by users who understand that the ipv6 switch should be set ti true in case of ipv6 addresses.

stratoula reviewed (commented):
CI complains here

<img width="1081" alt="image" src="https://github.com/elastic/kibana/assets/17003240/4e02ad64-1ed4-4542-b1c1-1c5f48696697">

Code seems fine to me and the behavior. Thanx for updating the docs. Let's make CI happy, I will test and review for a final time and I will merge. Thanx for your effort here!

ION28 commented:
@elasticmachine merge upstream

ION28 commented:
I finally figured out how to run at least some of the CI checks on my local machine since I can't get buildkite to work on my own account. I'm hoping there's no CI problems this time around finally 😅 It passes the jest tests for src/plugins/data and src/plugins/vis_default_editor/ on my box as well as scripts/type_check.js and others in that folder that I ran. 

Let me know if there's anything else you'd like me to address @stratoula . Thank you again for all of your feedback thus far!

stratoula commented:
/ci

stratoula commented:
@elasticmachine run elasticsearch-ci/docs

stratoula commented:
@elasticmachine run elasticsearch-ci/docs

delanni commented:
@elasticmachine run elasticsearch-ci/docs

delanni commented:
Admin-merging as per request, it seems we can't trigger the `buildkite/docs-build-pr Expected` check for this PR.
